### PR TITLE
Optimize stack manipulations in request flow

### PR DESCRIPTION
**Pull request by vladsud**
See issue https://github.com/microsoft/FluidFramework/issues/7394 for more details
UTs are required (will add).

### CODE REVIEW

**Comment by vladsud** [comment]
@josephliccini - FYI

**Comment by josephliccini** [comment]
Love this approach. Great way to balance not breaking the contract yet deferring the expensive browser API until absolutely needed.  Thank you for doing this!

File: packages/utils/telemetry-utils/src/errorLogging.ts [hunk 1]
```diff
@@ -166,17 +166,37 @@ export function normalizeError(
...
+ * There are scenarios where we generate error with stack, but error is handled in most cases and
+ * stack property is not accessed.
+ * For such cases it's better to not read stack property right away, but rather delay it until / if it's needed
+ * Some browsers will populate stack right away, others require throwing Error, so we do auto-detection on the fly.
+ * @returns Error object that has stack populated.
+ */
+ export function generateErrorWithStack(): Error {
+    const err = new Error("<<generated stack>>");
+
+    if (stackPopulatedOnCreation === undefined) {
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/errorLogging.ts:183]: Clever

File: packages/runtime/runtime-utils/src/dataStoreHelpers.ts [hunk 2]
```diff
@@ -44,31 +33,33 @@ export function exceptionToResponse(err: any): IResponse {
...
         mimeType: "text/plain",
         status,
         value: `${err}`,
-        stack: getStack(),
+        get stack() { return ((err?.stack) as (string|undefined)) ?? errWithStack.stack; },
     };
 }
 
-export function responseToException(response: IResponse, request: IRequest) {
+export function responseToException(response: IResponse, request: IRequest): Error {
```
Comment (markfields) [inline packages/runtime/runtime-utils/src/dataStoreHelpers.ts:51]: Is this function under test?  Code looks right but also the kind of diff where a bug could slip in...

File: packages/runtime/runtime-utils/src/dataStoreHelpers.ts [hunk 3]
```diff
@@ -44,31 +33,33 @@ export function exceptionToResponse(err: any): IResponse {
...
-    responseErr.code = response.status;
-    if (response.stack !== undefined) {
-        try {
-            // not clear if all browsers allow overwriting stack
-            responseErr.stack = response.stack;
-        } catch (err2) {}
-    }
-    return err;
+    const errWithStack = generateErrorWithStack();
+    const responseErr: Error & IResponseException = {
```
Comment (markfields) [inline packages/runtime/runtime-utils/src/dataStoreHelpers.ts:54]: Consider if we can make this IFluidError?  Best to use a class that extends LoggingError.  It's not easy if a suitable class doesn't already exist, I have thoughts here but haven't had time.

**Review by markfields — APPROVED** [review]
🎂 cool

File: packages/runtime/runtime-utils/src/dataStoreHelpers.ts (hunk 2 above)
Comment (markfields) [reply packages/runtime/runtime-utils/src/dataStoreHelpers.ts:51]: Oh I see your UT comment.  Good!

File: packages/runtime/runtime-utils/src/dataStoreHelpers.ts (hunk 3 above)
Comment (vladsud) [reply packages/runtime/runtime-utils/src/dataStoreHelpers.ts:54]: I'd prefer to move in small steps here. We can make such change later if desired

**Pull request merged**